<a href="https://colab.research.google.com/github/yuya1201/Digital-Advertising-HW/blob/master/Python_Exercise_3_Using_LASSO_Regression_to_Model_Fat_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import pandas
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoLarsCV
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, accuracy_score
import matplotlib.pyplot as plt

In [0]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
# load data
alldata=pd.read_csv('drive/My Drive/Colab Notebooks/finalmaster-ratios.csv')
alldata.head(3)

# Purchases  B01001001   B01001002  ...  B19001015  B19001016  B19001017
0           22     206252  469.226965  ...  36.440765  23.446284  21.197485
1            7      61399  486.538869  ...  33.000508  33.169741  24.792689
2            3      73170  489.859232  ...  20.125056  11.890525  16.537397

[3 rows x 190 columns]

In [0]:
# create a list of column names values
allvariablenames=list(alldata.columns.values)
listofallpredictors=allvariablenames[8:]

#load predictors into dataframe
predictors = alldata[listofallpredictors]  

#load target into dataframe
target = alldata['# Purchases']   

In [0]:
# split data into train and test sets, with 30% retained for test
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)    

In [0]:
# create model
model=LassoLarsCV(cv=10,precompute=False)
lassolarscv=model.fit(pred_train,tar_train)

In [0]:
# build coefficent chart
predictors_model=pd.DataFrame(listofallpredictors)
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)

['B01001014' 0.8558761066941788]
['B01001036' 2.5053482381631653]
['B01001037' 0.8892493223320962]
['B01001038' 1.5316387928880384]
['B02001005' 0.41252295298457853]
['B13014026' 0.48004105312075906]
['B13014027' 0.6978957445987839]
['B13016001' 875149895.329212]
['B19001017' 1.4834348068681533]


# **Question 1**

### **Line 1**
Change the list of all predictors into a data frame called predictors_model.
### **Line 2**
Add a column name for those predictors.
### **Line 3**
Add another column called coeff containing coefficient of each predictors.
### **Line 4**
Create a for loop to loop through each row of predictors_model.
### **Line 5**
Create a if statement to filter out rows with coeff that is greater than zero, under the for loop.
### **Line 6**
Print all predictors with their coefficients that are above zero.


# **Question 2**

**['B01001014' 0.8558761066941788]:**  
Males aged 40 to 44 years.  
One more male who aged 40 to 44 years makes 0.8558761066941788 unit more Bobo Bars sold.

**['B01001036' 2.5053482381631653]:**  
Females aged 30 to 34 years.  
One more female who aged 30 to 34 years makes 2.5053482381631653 units more Bobo Bars sold.

**['B01001037' 0.8892493223320962]:**  
Females aged 35 to 39 years.  
One more female who aged 35 to 39 years makes 0.8892493223320962 unit more Bobo Bars sold.

**['B01001038' 1.5316387928880384]:**  
Females aged 40 to 44 years.  
One more female who aged 40 to 44 years makes 1.5316387928880384 units more Bobo Bars sold.

**['B02001005' 0.41252295298457853]:**  
 Asian Alone.  
 One more Asian Alone makes .41252295298457853 unit more BObo Bars sold.

**['B13014026' 0.48004105312075906]:**  
Women 15 to 50 Years Who Had a Birth in the Past 12 Months, Unmarried (Never Married, Widowed and Divorced), with Bachelor's Degree.  
One more unmarried(never Married, widowed and divorced) woman who aged 15 to 50, had a birth in the past 12 months, and had a Bachelor's Degree makes  0.48004105312075906 unit more Bobo Bars sold.

**['B13014027' 0.6978957445987839]:**  
Women 15 to 50 Years Who Had a Birth in the Past 12 Months, Unmarried (Never Married, Widowed and Divorced), with Graduate or Professional Degree.  
One more unmarried(never Married, widowed and divorced) woman who aged 15 to 50, had a birth in the past 12 months, and had a Graduate or Professional Degree makes 0.6978957445987839 unit more Bobo Bars sold.

**['B13016001' 875149895.329212]:**  
All women 15 to 50 Years Who Had a Birth in the Past 12 Months.  
One more woman who aged 15 to 50 years and had a birth in th past 12 months makes 875149895.329212 units more Bobo Bars sold.

**['B19001017' 1.4834348068681533]:**   
Household with \$200,000 or more income in the Past 12 Months (In 2010 Inflation-Adjusted Dollars)  
One more household with \$200,000 or more income in the Past 12 Months makes 1.4834348068681533 units more Bobp Bars sold.

# **Question 3**

 If I had to report only two census variables to my boss that most steeply predicted sales, the first one would be women aged 15 to 50 years who had a birth in the past 12 months; the second would be females aged 30 to 34 years.

In [0]:
# train mean squared error
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
22025.491066757


In [0]:
# test mean squared error
test_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('testing data MSE')
print(test_error)

testing data MSE
41549.54803776253


# **Question 4**

The mean squared error of training data is 22025.491066757. The mean squared error of testing data is 41549.54803776253. They are not similar. Mean squared error measures how close a regression line close to data points. The smaller the mean squared error, the closer to find the line of best fit. Thus, training data is closer to the fitted line than the testing data is, which makes more sense because the model is build based on the training data.

In [0]:
# r squared - training
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.2400221219784492


In [0]:
# r squared - testing
rsquared_test=model.score(pred_test,tar_test)
print ('testing data R-square')
print(rsquared_test)

testing data R-square
0.1758628512005107


The R-squared for training data is 0.2400221219784492 and the R-squared for testing data is 0.1758628512005107. This means that the model could explain more variability of testing data than training data.

# **Question 5**

Census data does not predict sales well. It makes sense that training data has higher R-squared which means the model can explain more variability of training data. However, both mean squared error are not high enough to validate the model. Since MSEs are both not small enough, R-squared and MSE for both datasets are not sufficient enough to make good prediction.

In [0]:
# print model intercept
print("y interecept:")
print(model.intercept_)

y interecept:
22.19738813257551


# **Question 6**

The baseline sales number is 22.19738813257551. If value of every variable is zero, then there will be 22.19738813257551 Bobo Bars sold.

> Indented block

